docs/development/decision-records/2024-09-16-proposal-federated-catalog-distribution/README.md

```diff
@@ -0,0 +1,88 @@
+
+To enable the Federated Catalog flow, please [see this table](https://github.com/eclipse-tractusx/tractusx-edc/blob/75bdacbad43e2cad352204ea28a359c6aac7adea/docs/development/management-domains/README.md#enable-and-configure-the-crawler-subsystem).
+
+For its TargetNodeDirectory, the user is able to obtain the Connectors' URL's through the Discovery Service and store them in the new extension through its API. The API will allow to save a list of BPNLs (and Connectors' URL's if desired) and the `DiscoveryServiceRetrieverExtension` is responsible to retrieve the data and store it (in memory or in a database). The URL's can later be retrieved and crawled by the Federated Catalog.
+
+The retrieval of Connector URL's through the Discovery Service is enabled by the endpoint:
+```
+POST: /api/administration/connectors/discovery
+```
+In which, the body of the request can contain the BPNL's related with participants from which the catalogs want to be obtained. If no BPNL is provided (empty list) then all Connector URL's will be returned by the Discovery Service.
+Information regarding the related API can be found [here](https://catenax-ev.github.io/docs/standards/CX-0001-EDCDiscoveryAPI#22-api-specification).
+
+Some limitations of this TargetNodeDirectory solution are:
+- Each partner must have the BPNLs beforehand. If a new Partner registers and an existing partner would want their catalog, the BPNL (or Connector URL's) of the new partner must be obtained first;
+- Deal with the overhead an additional persistence store;
+- The usage of the Discovery Service requires a technical user account to access it (must be requested).
+
+
+As indicated, the new service would have own API capable of:
+
+#### Save BPNL's
+Request body would contain a list of BPNL's, allowing to store in bulk.
+```
+[POST] /api/target-nodes
+```
+Request Body Example
+```json
+[ "BPNL000000000001","BPNL000000000002" ]
+```
+
+#### Remove a stored BPNL
+BPNL to be remvoed is sent as a path param.
+```
+[DELETE] /api/target-nodes/{bpnl}
+```
+#### Retrieve BPNL's
+Get BPNL's (value and connectors associated with it).
+```
+[POST] /api/target-nodes/request
+```
+Request Body Example
+```json
+[ "BPNL000000000001","BPNL000000000002" ]
+```
+Response Example
+```json
+[
+    {
+        "bpn": "BPNL000000000001",
+        "connectorEndpoint": [
+            "https://connector1/api/v1/dsp"
+        ]
+    },
+    {
+        "bpn": "BPNL000000000002",
+        "connectorEndpoint": [
+            "https://connector2/api/v1/dsp",
```

Comment:
what does this mean? I know that connectors have multiple URLs in the DiscoveryService (which to me is a design flaw), but what does this mean for the TND? will there be multiple entries per BPN? If so, what does this mean for using the BPN as primary key? Will we need update the data model for the TargetNode?

Reply:
This is a very good point. 
Yes, the aim to have the BPNL as the pk will require the TargetNode targetUrl prop to be a list instead. Or consider a different pk and allowing the BPNL to not be unique. 
I would prefer to have the BPNL as pk and change the TargetNode.

Reply:
i would caution against using the BPN as primary key. Moving away from BPNs and toward DIDs fully is currently being discussed and this would cause problems with the data model later on. In addition, conflating business keys (such as BPNs) and database keys is somewhat of an antipattern.
There always is the possibility to wrap the `TargetNode` in another class and use that class as database entity.

Reply:
This is a good point, thank you.
For the time being the BPN will still be used despite the expected change in the future for the DID usage only. So storing BPN would still be ideal, I think, but maybe not as a db id.
Using a wrapper around the TargetNode is great, the only concern to me is to what would be the db key then. Not using the BPN and using the DID but still having the BPN in the wrapper class?

Reply:
database IDs should be synthetic, UUIDs usually are good for this. Just so you're aware, the `TargetNode.id` is used as `counterPartyId` in the `CatalogRequestMessage` later. That means, so long as we still use BPNs, `TargetNode.id` must contain the BPN.